{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### input and libs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "code_folding": [
     35,
     134
    ]
   },
   "outputs": [],
   "source": [
    "source(\"~/github/common_libs/mylibs.R\")\n",
    "source(\"~/github/common_libs/mylibs.R\")\n",
    "require(BSgenome.Mmusculus.UCSC.mm10)\n",
    "require(SummarizedExperiment)\n",
    "require(GenomicRanges)\n",
    "\n",
    "\n",
    "countInsertions <- function(query, fragments, by = \"RG\") {\n",
    "    # Count By Fragments Insertions\n",
    "    message(\"reading tags\")\n",
    "    tic()\n",
    "    inserts <- fread(fragments, col.names = c(\"chr\", \"start\", \"end\", \"RG\", \"score\", \n",
    "        \"strand\")) %>% dplyr::select(-score) %>% mutate(start = start + 1)\n",
    "    a = toc()\n",
    "    message(\"overlaping\")\n",
    "    inserts <- GRanges(seqnames = inserts$chr, ranges = IRanges(start = inserts$start, \n",
    "        end = inserts$end), strand = inserts$strand, RG = inserts$RG) %>% resize(width = 1, \n",
    "        fix = \"start\")\n",
    "    overlapDF <- DataFrame(findOverlaps(query, inserts, ignore.strand = TRUE, maxgap = -1L, \n",
    "        minoverlap = 0L, type = \"any\"))\n",
    "    overlapDF$name <- mcols(inserts)[overlapDF[, 2], by]\n",
    "    overlapTDF <- transform(overlapDF, id = match(name, unique(name)))\n",
    "    # Calculate Overlap Stats\n",
    "    inPeaks <- table(overlapDF$name)\n",
    "    total <- table(mcols(inserts)[, by])\n",
    "    total <- total[names(inPeaks)]\n",
    "    frip <- inPeaks/total\n",
    "    # Summarize\n",
    "    sparseM <- Matrix::sparseMatrix(i = overlapTDF[, 1], j = overlapTDF[, 4], x = rep(1, \n",
    "        nrow(overlapTDF)), dims = c(length(query), length(unique(overlapDF$name))))\n",
    "    colnames(sparseM) <- unique(overlapDF$name)\n",
    "    total <- total[colnames(sparseM)]\n",
    "    frip <- frip[colnames(sparseM)]\n",
    "    out <- list(counts = sparseM, frip = frip, total = total)\n",
    "    toc() - a\n",
    "    return(out)\n",
    "}\n",
    "\n",
    "\n",
    "extendedPeakSet <- function(df, BSgenome = NULL, extend = 250, blacklist = NULL, \n",
    "    nSummits = 1e+05) {\n",
    "    # Helper Functions\n",
    "    readSummits <- function(file) {\n",
    "        df <- fread(file, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\")) %>% \n",
    "            dplyr::select(1, 2, 3, 5)\n",
    "        return(GenomicRanges::makeGRangesFromDataFrame(df = df, keep.extra.columns = TRUE, \n",
    "            starts.in.df.are.0based = TRUE))\n",
    "    }\n",
    "    nonOverlappingGRanges <- function(gr, by = \"score\", decreasing = TRUE, verbose = FALSE) {\n",
    "        stopifnot(by %in% colnames(mcols(gr)))\n",
    "        clusterGRanges <- function(gr, filter = TRUE, by = \"score\", decreasing = TRUE) {\n",
    "            gr <- sort(sortSeqlevels(gr))\n",
    "            r <- GenomicRanges::reduce(gr, min.gapwidth = 0L, ignore.strand = TRUE)\n",
    "            o <- findOverlaps(gr, r)\n",
    "            mcols(gr)$cluster <- subjectHits(o)\n",
    "            gr <- gr[order(mcols(gr)[, by], decreasing = decreasing), ]\n",
    "            gr <- gr[!duplicated(mcols(gr)$cluster), ]\n",
    "            gr <- sort(sortSeqlevels(gr))\n",
    "            mcols(gr)$cluster <- NULL\n",
    "            return(gr)\n",
    "        }\n",
    "        if (verbose) {\n",
    "            message(\"Converging\", appendLF = FALSE)\n",
    "        }\n",
    "        i <- 0\n",
    "        gr_converge <- gr\n",
    "        while (length(gr_converge) > 0) {\n",
    "            if (verbose) {\n",
    "                message(\".\", appendLF = FALSE)\n",
    "            }\n",
    "            i <- i + 1\n",
    "            gr_selected <- clusterGRanges(gr = gr_converge, filter = TRUE, by = by, \n",
    "                decreasing = decreasing)\n",
    "            gr_converge <- subsetByOverlaps(gr_converge, gr_selected, invert = TRUE)  #blacklist selected gr\n",
    "            if (i == 1) {\n",
    "                # if i=1 then set gr_all to clustered\n",
    "                gr_all <- gr_selected\n",
    "            } else {\n",
    "                gr_all <- c(gr_all, gr_selected)\n",
    "            }\n",
    "        }\n",
    "        if (verbose) {\n",
    "            message(\"\\nSelected \", length(gr_all), \" from \", length(gr))\n",
    "        }\n",
    "        gr_all <- sort(sortSeqlevels(gr_all))\n",
    "        return(gr_all)\n",
    "    }\n",
    "    # Check-------\n",
    "    stopifnot(extend > 0)\n",
    "    stopifnot(\"samples\" %in% colnames(df))\n",
    "    stopifnot(\"groups\" %in% colnames(df))\n",
    "    stopifnot(\"summits\" %in% colnames(df))\n",
    "    stopifnot(!is.null(BSgenome))\n",
    "    stopifnot(all(apply(df, 1, function(x) {\n",
    "        file.exists(paste0(x[3]))\n",
    "    })))\n",
    "    #------------\n",
    "    # Deal with blacklist\n",
    "    if (is.null(blacklist)) {\n",
    "        blacklist <- GRanges()\n",
    "    } else if (is.character(blacklist)) {\n",
    "        blacklist <- rtracklayer::import.bed(blacklist)\n",
    "    }\n",
    "    stopifnot(inherits(blacklist, \"GenomicRanges\"))\n",
    "    #------------\n",
    "    # Time to do stuff\n",
    "    chromSizes <- GRanges(names(seqlengths(BSgenome)), IRanges(1, seqlengths(BSgenome)))\n",
    "    chromSizes <- GenomeInfoDb::keepStandardChromosomes(chromSizes, pruning.mode = \"coarse\")\n",
    "    groups <- unique(df$groups)\n",
    "    groupGRList <- GenomicRanges::GenomicRangesList(lapply(seq_along(groups), function(i) {\n",
    "        df_group = df[which(df$groups == groups[i]), ]\n",
    "        grList <- GenomicRanges::GenomicRangesList(lapply(paste0(df_group$summits), \n",
    "            function(x) {\n",
    "                extended_summits <- readSummits(x) %>% resize(., width = 2 * extend + \n",
    "                  1, fix = \"center\") %>% subsetByOverlaps(., chromSizes, type = \"within\") %>% \n",
    "                  subsetByOverlaps(., blacklist, invert = TRUE) %>% nonOverlappingGRanges(., \n",
    "                  by = \"score\", decreasing = TRUE)\n",
    "                extended_summits <- extended_summits[order(extended_summits$score, \n",
    "                  decreasing = TRUE)]\n",
    "                if (!is.null(nSummits)) {\n",
    "                  extended_summits <- head(extended_summits, nSummits)\n",
    "                }\n",
    "                mcols(extended_summits)$scoreQuantile <- trunc(rank(mcols(extended_summits)$score))/length(mcols(extended_summits)$score)\n",
    "                extended_summits\n",
    "            }))\n",
    "        # Non Overlapping\n",
    "        grNonOverlapping <- nonOverlappingGRanges(unlist(grList), by = \"scoreQuantile\", \n",
    "            decreasing = TRUE)\n",
    "        # Free Up Memory\n",
    "        remove(grList)\n",
    "        gc()\n",
    "        grNonOverlapping\n",
    "    }))\n",
    "    grFinal <- nonOverlappingGRanges(unlist(groupGRList), by = \"scoreQuantile\", decreasing = TRUE)\n",
    "    grFinal <- sort(sortSeqlevels(grFinal))\n",
    "    return(grFinal)\n",
    "}\n",
    "\n",
    "split_write_per_lib <- function(lib = \"JYH_854_1_2\", clust_res, data_dir = \"~/data/outputs/snATACj/\", \n",
    "    save_dir = \"~/scratch/outputs_snATACj/immune_allCons/rd1_clust_bed/tags/\", verbose = F) {\n",
    "    ### split reads\n",
    "    split_reads <- function() {\n",
    "        ### split reads for lib's tagalign\n",
    "        tic()\n",
    "        message(\"read tagalign file...\")\n",
    "        frag_lib <- fread(paste0(\"zcat \", data_dir, lib, \"/\", lib, \".filt.nodup.tn5.tagAlign.gz\"), \n",
    "            header = F)\n",
    "        t1 = toc()\n",
    "        message(\"splitting reads\")\n",
    "        frag_lib <- frag_lib %>% inner_join(clust_res, by = c(V4 = \"index\")) %>% \n",
    "            setDT\n",
    "        frag_lib_split = split(frag_lib, by = c(\"leiden\"), keep.by = F)\n",
    "        # rm(frag_lib)\n",
    "        toc() - t1\n",
    "        return(frag_lib_split)\n",
    "    }\n",
    "    ### save bed files and peak calling\n",
    "    save_tag <- function(frag_lib_split) {\n",
    "        if (!dir.exists(save_dir)) \n",
    "            dir.create(save_dir, recursive = T)\n",
    "        sapply(names(frag_lib_split), function(i) {\n",
    "            if (verbose) \n",
    "                message(\"write \", lib, \" clust \", i)\n",
    "            fwrite(frag_lib_split[[i]] %>% dplyr::select(starts_with(\"V\")), file = paste0(save_dir, \n",
    "                \"c\", i, \".tagAlign\"), append = T, sep = \"\\t\", col.names = F)\n",
    "        })\n",
    "    }\n",
    "    ## main\n",
    "    save_tag(split_reads())\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "### split reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "clust_res = \"./cluster_res.csv\"\n",
    "clust_res <- fread(clust_res) %>% dplyr::select(-starts_with(\"umap\"))\n",
    "clust_res$lib <- sub(\"_[ATCG]+$\", \"\", clust_res$index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "running XH_206_1_2\n",
      "read tagalign file...\n",
      "Warning message in fread(paste0(\"zcat \", data_dir, lib, \"/\", lib, \".filt.nodup.tn5.tagAlign.gz\"), :\n",
      "\"Previous fread() session was not cleaned up properly. Cleaned up ok at the beginning of this fread() call.\""
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "elapsed time is 177.008000 seconds \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "splitting reads\n"
     ]
    }
   ],
   "source": [
    "for (l in (clust_res$lib %>% unique)[7:9]) {\n",
    "    message(\"running \", l)\n",
    "    split_write_per_lib(lib = l, clust_res = clust_res)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "### call peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "code_folding": [
     7
    ],
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol>\n",
       "\t<li></li>\n",
       "\t<li></li>\n",
       "\t<li></li>\n",
       "\t<li></li>\n",
       "\t<li></li>\n",
       "\t<li></li>\n",
       "\t<li></li>\n",
       "\t<li></li>\n",
       "\t<li></li>\n",
       "\t<li></li>\n",
       "\t<li></li>\n",
       "\t<li></li>\n",
       "\t<li></li>\n",
       "\t<li></li>\n",
       "\t<li></li>\n",
       "\t<li></li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate}\n",
       "\\item \n",
       "\\item \n",
       "\\item \n",
       "\\item \n",
       "\\item \n",
       "\\item \n",
       "\\item \n",
       "\\item \n",
       "\\item \n",
       "\\item \n",
       "\\item \n",
       "\\item \n",
       "\\item \n",
       "\\item \n",
       "\\item \n",
       "\\item \n",
       "\\end{enumerate}\n"
      ],
      "text/markdown": [
       "1. \n",
       "2. \n",
       "3. \n",
       "4. \n",
       "5. \n",
       "6. \n",
       "7. \n",
       "8. \n",
       "9. \n",
       "10. \n",
       "11. \n",
       "12. \n",
       "13. \n",
       "14. \n",
       "15. \n",
       "16. \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[[1]]\n",
       "character(0)\n",
       "\n",
       "[[2]]\n",
       "character(0)\n",
       "\n",
       "[[3]]\n",
       "character(0)\n",
       "\n",
       "[[4]]\n",
       "character(0)\n",
       "\n",
       "[[5]]\n",
       "character(0)\n",
       "\n",
       "[[6]]\n",
       "character(0)\n",
       "\n",
       "[[7]]\n",
       "character(0)\n",
       "\n",
       "[[8]]\n",
       "character(0)\n",
       "\n",
       "[[9]]\n",
       "character(0)\n",
       "\n",
       "[[10]]\n",
       "character(0)\n",
       "\n",
       "[[11]]\n",
       "character(0)\n",
       "\n",
       "[[12]]\n",
       "character(0)\n",
       "\n",
       "[[13]]\n",
       "character(0)\n",
       "\n",
       "[[14]]\n",
       "character(0)\n",
       "\n",
       "[[15]]\n",
       "character(0)\n",
       "\n",
       "[[16]]\n",
       "character(0)\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dir_peaks <- \"~/scratch/outputs_snATACj/immune_allCons/rd1_clust_bed/peaks/\"\n",
    "if (!dir.exists(save_dir)) dir.create(save_dir, recursive = T)\n",
    "\n",
    "method <- \"q\"\n",
    "cutoff <- 0.05\n",
    "shift <- -75\n",
    "extsize <- 150\n",
    "genome_size <- 1.87e+09\n",
    "n_clust <- clust_res %>% pull(leiden) %>% unique %>% length\n",
    "mclapply(1:n_clust - 1, function(j) {\n",
    "    message(sprintf(\"%s of %s\", j, n_clust))\n",
    "    cluster_tagj <- paste0(dir_peaks, \"c\", j, \".tagAlign\")\n",
    "    cmdPeaks <- sprintf(\"macs2 callpeak -g %s --name %s --treatment %s --outdir %s --format BED --nomodel --call-summits --nolambda --keep-dup all\", \n",
    "        genome_size, paste0(\"c\", j), cluster_tagj, dir_peaks)\n",
    "    if (!is.null(shift) & !is.null(extsize)) {\n",
    "        cmdPeaks <- sprintf(\"%s --shift %s --extsize %s\", cmdPeaks, shift, extsize)\n",
    "    }\n",
    "    if (tolower(method) == \"p\") {\n",
    "        cmdPeaks <- sprintf(\"%s -p %s\", cmdPeaks, cutoff)\n",
    "    } else {\n",
    "        cmdPeaks <- sprintf(\"%s -q %s\", cmdPeaks, cutoff)\n",
    "    }\n",
    "    message(\"Running Macs2...\")\n",
    "    message(cmdPeaks)\n",
    "    system(cmdPeaks, intern = TRUE)\n",
    "}, mc.cores = 4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make Non-Overlapping Peak Set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df <- data.frame(samples = gsub(\"\\\\_summits.bed\", \"\", list.files(dir_peaks, pattern = \"\\\\_summits.bed\", \n",
    "    full.names = FALSE)), groups = \"scATAC\", summits = list.files(dir_peaks, pattern = \"\\\\_summits.bed\", \n",
    "    full.names = TRUE))\n",
    "\n",
    "\n",
    "unionPeaks <- extendedPeakSet(df = df, BSgenome = genome, extend = 250, blacklist = \"~/data/GENOME/mm10/mm10.blacklist.bed\", \n",
    "    nSummits = 2e+05)\n",
    "unionPeaks <- unionPeaks[seqnames(unionPeaks) %in% paste0(\"chr\", c(1:19, \"X\"))]\n",
    "unionPeaks <- keepSeqlevels(unionPeaks, paste0(\"chr\", c(1:19, \"X\")))\n",
    "\n",
    "# Create Counts list\n",
    "fragmentFiles <- list.files(dir_peaks, pattern = \"\\\\.tagAlign\", full.names = T)\n",
    "countsPeaksList <- lapply(seq_along(fragmentFiles), function(i) {\n",
    "    message(sprintf(\"%s of %s\", i, length(fragmentFiles)))\n",
    "    gc()\n",
    "    countInsertions(unionPeaks, fragmentFiles[i], by = \"RG\")\n",
    "})\n",
    "\n",
    "# CountsMatrix\n",
    "mat <- lapply(countsPeaksList, function(x) x[[1]]) %>% Reduce(\"cbind\", .)\n",
    "frip <- lapply(countsPeaksList, function(x) x[[2]]) %>% unlist\n",
    "total <- lapply(countsPeaksList, function(x) x[[3]]) %>% unlist\n",
    "\n",
    "se <- SummarizedExperiment(assays = SimpleList(counts = mat), rowRanges = unionPeaks)\n",
    "rownames(se) <- paste(seqnames(se), start(se), end(se), sep = \"_\")\n",
    "colData(se)$FRIP <- frip\n",
    "colData(se)$uniqueFrags <- total/2\n",
    "saveRDS(se, \"~/scratch/outputs_snATACj/immune_allCons/rd1_clust_bed/peak_cell_se.rds\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
