#!/bin/bash
#PBS -q hotel
#PBS -N snATAC_pre_j
#PBS -l nodes=1:ppn=16
#PBS -l walltime=16:00:00
#PBS -V
#PBS -m abe
#PBS -A epigen-group

source activate scanpy
export PATH=/projects/ps-epigen/software/snATAC_josh/:$PATH;


############################################################
# Input
############################################################
# sample=$1 #JYH_xxx, -v
samplenames=(`cat $samples`)
sample=${samplenames[${PBS_ARRAYID}]} #index start from 0


work_dir="/home/zhc268/scratch/outputs_snATACj"
output_dir=${work_dir}/${sample} #/oasis/tscc/scratch/zhc268/projects/evans/190320_immune_cell/Josh/${sample}
mkdir -p $output_dir
genome=mm10
chromsize="/projects/ps-epigen/GENOME/${genome}/${genome}.chrom.sizes"
blacklist="/projects/ps-epigen/GENOME/${genome}/${genome}.blacklist.bed"
input_bam="/home/zhc268/data/outputs/10xATAC/${sample}/outs/possorted_bam.bam"
prom="/projects/ps-epigen/GENOME/${genome}/gencode.vM17.protein_coding.gene.prom.2kb.bed"
log="${output_dir}/preprocess_log.txt"
threads=$PBS_NP
if [ -d /home/zhc268/data/outputs/peaks/${sample}/ ]
then
    npeak_file=$(find /home/zhc268/data/outputs/peaks/${sample}/ -name "*filt*narrow*.gz")
else
    npeak_file=""
fi

[[ ! -z $npeak_file ]]  && peak_args="--peak-file $npeak_file"
[[ -z $npeak_file ]]  && peak_args=""

### 1 poreprocess
# --skip-convert
cmd="time snATAC_pipeline_10X.py -o $output_dir  -b $input_bam -n ${sample} --shift -75 --extsize 150 \
     --chrom-sizes $chromsize --blacklist-file $blacklist -t $threads \
     $peak_args  -m 1 \
     --promoter-file $prom 1> $log 2>&1"

echo -e $cmd

time eval $cmd


############################################################
# transfer 
############################################################
rsync -avz $output_dir --exclude '*md.bam*' --exclude "*compiled.filt.bam" /home/zhc268/data/outputs/snATACj/
